Features/dereferencing support with circular references

repocho opened the pull request:
Fixes #179 

Finally I limited the circular references up to 500 by default which in my case is not throwing errors. 

If there is a harder limit of nodejs you can set the maxDepth like this example:

`raml.loadFile('raml-files/circular.raml', {
      dereferenceSchemas: { maxDepth: 25}
})`

repocho commented:
I've updated the travis nodejs version to the latest LTS version 4.4.4

dmartinezg commented on test/specs/dereferenceCircular.js:
```diff
@@ -0,0 +1,56 @@
+
+  chai.should();
+}
+var schema = "{\"$schema\":\"http://json-schema.org/draft-04/schema#\",\"definitions\":{\"node\":{\"type\":\"object\",\"properties\":{\"id\":{\"type\":\"string\",\"format\":\"id\"},\"label\":{\"type\":\"string\"},\"children\":{\"type\":\"array\",\"items\":{\"type\":\"object\",\"required\":[\"id\",\"label\",\"children\"]}}}}},\"type\":\"array\"}";
```
Using single quotes in the string, would help with all those backslashes

repocho replied:
Perfect, I prefer single quotes. I was following the same as the dereferencing tests.

dmartinezg commented on src/util.coffee:
```diff
@@ -32,3 +32,25 @@
 
 @NON_PRINTABLE = /[^\x09\x0A\x0D\x20-\x7E\x85\xA0-\uD7FF\uE000-\uFFFD]/
+
+@stringifyJSONCircularDepth = (json, depth = 1000) ->
```
I wonder when would a cycle ever be broken, why would counting to 500 be more effective than counting to 2?

What we do with includes is keep a list of all the URIs for the include, when a single one in the path is repeated, we abort with a circular reference error.

repocho replied:
> I wonder when would a cycle ever be broken, why would counting to 500 be more effective than counting to 2?

The depth is the max depth of a circular reference, you can check both files:
https://github.com/repocho/raml-js-parser/blob/deb2ef9ec8b2a797a6e4d7f17267f169b07c1782/test/specs/dereferenceCircularDepth1.js
https://github.com/repocho/raml-js-parser/blob/deb2ef9ec8b2a797a6e4d7f17267f169b07c1782/test/specs/dereferenceCircular.js
And check that the generated schema have an aditional level depth with `maxDepth=1` .

_About the includes you meant in the raml definition ?_ The objective of this PR is to avoid the circular exception of JSON.stringify of a dereference schema definition with circular dependencies of $ref.

dmartinezg replied:
What I meant is, why have an arbitrary length at all, the first time you detect a cycle, you should abort (just like we do in other parts)

repocho replied:
With this approach you are letting the users to choose it.
I mean, if this is the default approach we can put `maxDepth=0` by default. But for `raml-mocker` library I think is better to have high depth in order to have more diversity of generated jsons.

derrickburns replied:
IMHO, max depth is a hack.  Circular dependency checking is simple.  

At the same time, no human generated circular dependencies will have depth > 500.  So, it really doesn't matter.

dmartinezg replied:
My sentiment exactly @derrickburns, I think that the parser should just abort on any circular reference detected.

repocho replied:
I don't agree ! With a schema like this one you will abort the dereferencing parsing?
```json
{
    "$schema": "http://json-schema.org/schema",
    "definitions": {
        "node": {
            "type": "object",
            "properties": {
                "id": {
                    "type": "string",
                    "format": "id"
                },
                "label": {
                    "type": "string"
                },
                "children": {
                    "type": "array",
                    "items": {
                        "$ref": "#/definitions/node"
                    }
                }
            },
            "required": ["id", "label", "children"]
        }
    },
    "items": {
        "$ref": "#/definitions/node"
    },
    "type": "array"
}
```
If I want generate mocks from this schema how should I do it?

repocho replied:
@derrickburns @dmartinezg If you want to abort circulars dependencies by default. It's not a problem, but you should allow the possibility to parse circular schemas with dereference and a max depth.

Is this aligned with your thoughts ?

repocho replied:
@derrickburns @dmartinezg Any decision here?

dmartinezg replied:
Hey @repocho, you are absolutely right, this is the only way to handle recursive data structures. 1000 seems a little high still.

If you have signed this: https://api-notebook.anypoint.mulesoft.com/notebooks#bc1cf75a0284268407e4

I'll merge and publish. Thanks for your contrib.

repocho replied:
Hi @dmartinezg!

Actually the util function default is 1000 but the use in raml.coffee default is 500. I thought that value was more realistic (Ref: https://github.com/raml-org/raml-js-parser/pull/180/files#diff-fe0283a18f459b13627ffad0f22b8eb5R161)

I have also signed the contributor agreement.

Thanks.

repocho commented:
@dmartinezg Travis is failing because the used version of phantomjs does not allow `Object.assign`. Should I update the `grunt-mocha-phantomjs` to the latest version `3.0.0`?

repocho commented:
@dmartinezg I updated grunt-mocha-phantomjs library but the error was still there due to the limitation of phamtomjs engine with es6 features.
Finally I've added a shim library for `Object.assign` to avoid problems to people using this inside an old browser.

If you have more concerns about this PR please let me know in order to be able to merge it as soon as possible.

Thanks !

dmartinezg commented:
Can you squash commits so I can merge from here? I don't want to close this PR and steal your contribution :)

repocho commented:
Done ! Thanks.
